### ISSUE TITLE
Crash when trying to go to definition: undefined variable: g:phpactorProjectRootPatterns

### ISSUE THREAD

**Issue by greg0ire** [phpactor/phpactor#947]
phpactor crashes when I try to use `GotoDefinition` on the `DoctrineCommand` class here: https://github.com/doctrine/DoctrineFixturesBundle/blob/master/Command/LoadDataFixturesDoctrineCommand.php#L30

```
Error detected while processing function phpactor#GotoDefinition[1]..phpactor#_GotoDefinitionTarget[1]..phpactor#rpc:
line    6:
E121: Undefined variable: g:phpactorProjectRootPatterns
E116: Invalid arguments for function s:searchDirectoryUpwardForRootPatterns( fnamemodify(phpactor#_path(), ':h'), g:phpactorProjectRootPatterns, g:phpactorInitialCwd)
line   12:
E121: Undefined variable: g:phpactorPhpBin
line   14:
E121: Undefined variable: l:cmd
E116: Invalid arguments for function system(l:cmd, json_encode(request))
Error detected while processing function phpactor#GotoDefinition[1]..phpactor#_GotoDefinitionTarget[1]..phpactor#rpc:
line   20:
E605: Exception not caught: Could not parse response from Phpactor: Vim(let):E121: Undefined variable: result
```

**Comment by elythyr** [comment]
Hi, on master or develop ?
You got the error every time ? Others RPC call works as expected ?
Have you try to turn vim off and on again ? :) (but seriously in case a previous action would have unset a variable or something)

**Comment by greg0ire** [comment]
I'm on develop (and thought I was on master)
I have the error every time, even if that's the first action on my vim session. Not sure what things are RPC calls and what aren't, but `Transform` gives the same error, for instance.

**Comment by elythyr** [comment]
That's strange, I don't have any issue. I just updated my develop branch.
Can you try to, just in case something was fixed in the mid while?

The only thing I can think of is that somehow your filetype isn't `php`, so while your are in the buffer can you gave the result for:
```vim
:echo &ft
:echo g:phpactorProjectRootPatterns
```

**Comment by dantleech** [comment]
This is strange, maybe I'm not understanding something about how VIM is working (at least with standard VIM).

The variable is defined [here](https://github.com/phpactor/phpactor/blob/4c143024e2290ff75b412bfde64f4ea40eea98d0/ftplugin/php.vim#L51). Which should load when you open a PHP file?

**Comment by dantleech** [comment]
Also `:PhpactorGotoDefinition` seems to work fine on develop with:

```
:version
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Mar 24 2020 03:21:40)
```

**Comment by elythyr** [comment]
The `ftplugin/php.vim` is nothing new for vim so it should be handled even with an old version.
The only reason for me would be that phpactor is not in the `runtimepath` option but in this case he won't have access to any functionality, working or not.

**Comment by dantleech** [comment]
Could also be something to do with `Plug` - as I think @greg0ire is using the native VIM package manager, and the other issue was using  Dein.

**Comment by greg0ire** [comment]
I updated before reporting the bug, and am using 4c14302 right now.
`&ft` gives `php`, and the other command `E121: Undefined variable: g:phpactorProjectRootPatterns`

Using `VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Mar  3 2020 00:00:00)`

**Comment by elythyr** [comment]
OK, I think I managed to reproduce it with only phpactor as plugin on Vim

**Comment by elythyr** [comment]
If I remember correctly your vimrc @greg0ire you have something like:
```vim
autocmd FileType php packadd phpactor
```

As a temporary fix, remove this line from your vimrc and move phpactor from the `opt` directory into the `start` directory.
Somehow when using `packadd` the `ftplugin` directory appears as not being loaded.

**Comment by elythyr** [comment]
After looking quickly I think it's the expected behavior for both vim and neovim.
From `:h pack-add`:
> Filetype plugins should go under "pack/*/start", so that they are always
found.  Unless you have more than one plugin for a file type and want to

Phpactor was not a filetype plugin to begin with but it became one recently.
Which means that making it an optional package is not necessary (it's even a problem apparently) because it will be loaded only for the `php` filetype.

For those who are looking for an explanation, I think this does not works because the plugin is added after the `FileType` event is triggered.
So it's vim already look into the `runtimepath` for any `ftplugin/php.vim` or `ftplugin/php/*.vim` files.
Because if you call `packadd! phpactor` in your `vimrc` then it would work as expected.

**Comment by greg0ire** [comment]
> As a temporary fix, remove this line from your vimrc and move phpactor from the opt directory into the start directory.

That works indeed, thanks! cc @kristijanhusak, this might be of some interest to you since phpactor is used as a good example of an optional package in the README of vim-packager.

**Comment by dantleech** [comment]
Can we fix / mitigate this issue on Phpactor side?

**Comment by kristijanhusak** [comment]
@dantleech I guess moving variables to a plugin/phpactor.vim file would do the trick. Even if those things are loaded its' not much, 10 lines maybe.

**Comment by elythyr** [comment]
> @dantleech I guess moving variables to a plugin/phpactor.vim file would do the trick. Even if those things are loaded its' not much, 10 lines maybe.

I don't really agree on that, for me the problem is not within phpactor but how the vim package feature is used.

A filetype plugin could do a lot more than initializing variables:
* creates user events
* defines auto commands
* defines commands
* ...

It only make sense for a plugin like phpactor to be a filetype plugin.
Let's assume that phpactor provides mappings (for the sake of the argument).
In this scenario `<C-]>` is mapped to the command `PhpactorGotoDefinition`.
If this is not defined on a filetype plugin with the `<buffer>` argument then the mapping will always be defined.
Event with `au FileType php packadd phpactor`, the first buffer matching the filetype will be loaded for all buffers.
So we need to use auto commands based on filetype in phpactor to initialize the mappings...

Bottom line, for me a plugin should not take into account that some people might want to use it a certain way and add tricks in order to works in a scenario that is discouraged by the official documentation.
But we should update the install/configuration documentation to cover the case of optional packages and warn about this issue.

**Comment by kristijanhusak** [comment]
@elythyr I agree on your point. I just gave a possible solution to @dantleech question. I'm not sure if it will work.
My guess is that if we define only variables in the plugin/phpactor.vim file, and leave everything else in ftplugin file, it should work. Error that happens says that variable is missing, that doesn't necessarily means that other things are broken.

**Comment by elythyr** [comment]
> Error that happens says that variable is missing, that doesn't necessarily means that other things are broken.

I was actually thinking ahead, @dantleech wants phpactor vim integration to be the same for everyone and to not depends on outside dependencies.
Which is something I can understand, but I still like to be able to use great tools like fzf and stuff in order to make the editor more friendly.
In order to achieve that we might need to add some custom events to the plugin in order to gave the opportunity to other plugins to plug onto them and dynamically change the phpactor configuration.

Therefore I think we don't need to fix anything on this side since the bug is only on develop and it would possibly be limiting our options

**Comment by dantleech** [comment]
I think there is another factor here:

> I'm using :call phpactor#GotoDefinition

So that is now "deprecated" and everybody should use the command equivalent, `:PhpactorGotoDefinition` - which is only defined with the FT Plugin.

The documentation now reflects this, but I guess lots of people are still using the functions directly...

**Comment by elythyr** [comment]
> The documentation now reflects this, but I guess lots of people are still using the functions directly...

That's the problem of not having `<Plug>(PhpactorGotoDefinition)` mappings (expect if I missed that we have added them.
This just a nice way to define the interface we want to expose and allow us to switch what their are using in the background.

> I think there is another factor here

You're right it should not be used but it does not causes the bug.

**Comment by dantleech** [comment]
> That's the problem of not having <Plug>(PhpactorGotoDefinition)

Yes we should add that :) it would solve the issue by just making those functions unavailable?

**Comment by elythyr** [comment]
No, the functions would still be available and we should keep it that way since they just got deprecated.
I don't how many version you want to wait until introducing a BC break.

From `:h <Plug>`:
> The special key name "<Plug>" can be used for an internal mapping, which is
not to be matched with any key sequence.


A little bit clearer with `:h using-<Plug>`:
> <Plug> is a special code that a typed key will never produce.

So the idea is that you create a mapping that will never conflict with any one, as long as the name is unique enough of course.
Doing so the user just have to do:
```vim
nmap <buffer> <silent> <C-]> <Plug>(PhpactorGotoDefinition)
```

We will have to check the recommendations for the parenthesis because you see plugins with and without them.

**Comment by przepompownia** [comment]
`<Plug>()` mapping can be understood as an abstraction between the action assigned to it and any real mapping (to be made by the user like in above comment). It allows to provide a set of mappings reasonable for plugins without polluting user's custom mappings. Correctly named abstract mappings can then be easier to search in help document and listed (`:Maps` in fzf.vim).

**Comment by dantleech** [comment]
Reverted this change now for a simpler implementation
